state-reconstruct-fetcher/src/types/v3.rs

```diff
@@ -141,9 +148,113 @@ fn parse_pubdata_from_calldata(
 
     // Parse compressed state diffs.
     // NOTE: Is this correct? Ignoring the last 32 bytes?
-    let end_point = bytes.len() - 32;
-    let mut state_diffs = parse_compressed_state_diffs(&bytes[..end_point], pointer)?;
+    let diff_bytes = if shorten {
+        let end_point = bytes.len() - 32;
+        &bytes[..end_point]
+    } else {
+        bytes
+    };
+    let mut state_diffs = parse_compressed_state_diffs(diff_bytes, pointer)?;
     l2_to_l1_pubdata.append(&mut state_diffs);
 
     Ok(l2_to_l1_pubdata)
 }
+
+async fn parse_pubdata_from_blobs(
+    bytes: &[u8],
+    pointer: &mut usize,
+    client: &reqwest::Client,
+    blobs_url: &str,
+) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
+    let mut l = bytes.len() - *pointer;
+    let mut blobs = Vec::new();
+    while *pointer < l {
+        let pubdata_commitment = &bytes[*pointer..*pointer + PUBDATA_COMMITMENT_SIZE];
+        let blob = get_blob(&pubdata_commitment[48..96], client, blobs_url).await?;
+        let mut blob_bytes = ethereum_4844_data_into_zksync_pubdata(&blob);
+        blobs.append(&mut blob_bytes);
+        *pointer += PUBDATA_COMMITMENT_SIZE;
+    }
+
+    l = blobs.len();
+    while l > 0 && blobs[l - 1] == 0u8 {
+        l -= 1;
+    }
+
+    let blobs_view = &blobs[..l];
+    let mut pointer = 0;
+    parse_pubdata_from_calldata(blobs_view, &mut pointer, false)
+}
+
+async fn get_blob(
+    kzg_commitment: &[u8],
+    client: &reqwest::Client,
+    blobs_url: &str,
+) -> Result<Vec<u8>, ParseError> {
+    let url = format!("{}0x{}", blobs_url, hex::encode(kzg_commitment));
+    for attempt in 1..=MAX_RETRIES {
+        match client.get(url.clone()).send().await {
+            Ok(response) => match response.text().await {
+                Ok(text) => match get_blob_data(&text) {
+                    Ok(data) => {
```

Comment:
I want to keep the logging - is it possible to do that with the `?` operator?

Reply:
Not at the same level, but of course it's possible to still have logger at the caller level (by matching). If you _absolutely_ need logging, I'd recommend sticking to `let else`.

Though, in my personal opinion, a warning trace saying `data is not string` followed by an error trace saying the exact same thing (just without the source string) seems a bit superfluous. I'd advise to remove the `tracing::warn!`s (from `get_blob_data`) and just keep the errors since they get traced at the caller level anyways.

Reply:
But, the source that couldn't be parsed is the most important thing for debugging - if we don't log it, we won't even know whether it's changed... There's a lot of logging here because there was (and is) a lot of problems parsing blob data...

Reply:
Yeah, but right now we're logging a warning, and then we also return an error that gets caught and logged by the caller, so essentially we're logging twice for every single one of those errors, right? I think it makes more sense to instead pass in the source string in the error and avoid logging those warnings.

Reply:
Well yes, it's logged twice - like in lots of other places, e.g. in `retry_call` and its callers... But yes, the error data can be passed up before logging it...